src/main/scala/org/constellation/consensus/RoundManager.scala

```diff
@@ -267,44 +284,42 @@ object RoundManager {
-
-    val transactions =
-      dao.transactionService
-        .pullForConsensus(dao.minCheckpointFormationThreshold)
-        .unsafeRunSync()
-    if (transactions.nonEmpty) {
-      dao
-        .pullTips(dao.readyFacilitatorsAsync.unsafeRunSync())
-        .map { tips =>
-          val messages = dao.threadSafeMessageMemPool.pull().getOrElse(Seq()) // TODO: Choose more than one tx and light peers
-          val firstTx = transactions.headOption
-          val lightPeers =
-            if (firstTx.isDefined && dao.readyPeers(NodeType.Light).unsafeRunSync().nonEmpty) {
-              Set(
-                dao
-                  .readyPeers(NodeType.Light)
-                  .unsafeRunSync()
-                  .minBy(p => Distance.calculate(firstTx.get.transaction.baseHash, p._1))
-                  ._2
-              )
-            } else Set[PeerData]()
-          val allFacilitators = tips._2.values.map(_.peerMetadata.id).toSet ++ Set(dao.id)
-          (
-            RoundData(
-              generateRoundId,
-              tips._2.values.toSet,
-              lightPeers,
-              FacilitatorId(dao.id),
-              transactions.map(_.transaction),
-              tips._1,
-              messages
-            ),
-            getArbitraryTransactionsWithDistance(allFacilitators, dao).filter(t => t._2 == 1),
-            getArbitraryMessagesWithDistance(allFacilitators, dao).filter(t => t._2 == 1)
-          )
-        }
-    } else None
+  ): Future[Option[(RoundData, Seq[(Transaction, Int)], Seq[(ChannelMessage, Int)])]] = {
+
+    val task = for {
+      transactions <- dao.transactionService.pullForConsensus(dao.minCheckpointFormationThreshold)
+      _ <- if (transactions.isEmpty) IO.raiseError[Unit](NoTransactionsForConsensus) else IO.unit
+      facilitators <- dao.readyFacilitatorsAsync
+      tips = dao.pullTips(facilitators)
+      _ <- if (tips.isEmpty) IO.raiseError[Unit](NoTransactionsForConsensus) else IO.unit
+      messages = dao.threadSafeMessageMemPool.pull().getOrElse(Seq())
+      lightNodes <- dao.readyPeers(NodeType.Light)
+      lightPeers = if (lightNodes.isEmpty) Set.empty[PeerData]
+      else
+        Set(lightNodes.minBy(p => Distance.calculate(transactions.head.transaction.baseHash, p._1))._2) // TODO: Choose more than one tx and light peers
+      allFacilitators = tips.get.peers.values.map(_.peerMetadata.id).toSet ++ Set(dao.id)
+      roundData = (
+        RoundData(
+          generateRoundId,
+          tips.get.peers.values.toSet,
+          lightPeers,
+          FacilitatorId(dao.id),
+          transactions.map(_.transaction),
+          tips.get.tipSoe,
+          messages
+        ),
+        getArbitraryTransactionsWithDistance(allFacilitators, dao).filter(t => t._2 == 1),
+        getArbitraryMessagesWithDistance(allFacilitators, dao).filter(t => t._2 == 1)
+      )
+
+    } yield roundData.some
+
+    task.handleError {
+      case NoTransactionsForConsensus =>
+        println("catched")
```

Comment:
@marcinwadon You didn't spotted this one

Reply:
I've seen that